pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs

```diff
@@ -254,19 +255,28 @@ startServer
   -> Net.Socket
   -> Maybe W.Port
   -> TVar E.LiveReqs
+  -> IO ()
   -> RIO e ()
-startServer typ hos por sok red vLive = do
+startServer typ hos por sok red vLive onFatal = do
   envir <- ask
 
   let host = case hos of
         SHLocalhost -> "127.0.0.1"
         SHAnyHostOk -> "*"
 
+  let handler r e
+        | Just (IOError {ioe_type = ResourceExhausted}) <- fromException e
+        = runRIO envir $ do
+            logError (displayShow e)
+            io onFatal
+        | otherwise = W.defaultOnException r e
```

Comment:
What does the default exception do? Is that what's currently printing and then hanging?

Reply:
The default handler is found at https://hackage.haskell.org/package/warp-3.3.14/docs/src/Network.Wai.Handler.Warp.Settings.html#defaultOnException. It either prints or does not print the exception depending on the type, then discards it. Accordingly, it's the origin of the printout @lycurgus reported. It's not the source of the hang; that's because warp gets into a bad state when it runs out of file descriptors; it's not a generic thing that happens on any exception we might encounter here, nor does "the exception handler hang" as you describe it.

Reply:
Do you have any indication that printing/ignoring exceptions is the right default behavior here? It seems like it'd be better to whitelist non-fatal exceptions and otherwise `abort()`.

Reply:
Yeah, I do. I looked for examples of exception handling behavior in major open source codebases using warp. Take a look at:

- https://github.com/hasura/graphql-engine/blob/master/server/src-lib/Hasura/App.hs#L583
- https://github.com/PostgREST/postgrest/blob/main/main/Main.hs#L89

Both of these major, legitimate applications *fail to modify the default exception handler at all*, so they definitely don't `abort()` on any exception. 

Now the comment at https://hackage.haskell.org/package/warp-3.3.14/docs/src/Network.Wai.Handler.Warp.Settings.html#defaultShouldDisplayException says

> The goal is to hide exceptions which occur under the normal course of the web server running.

So perhaps the list found in that function will serve as a suitable whitelist? But in fact, the stderr printouts reported in https://github.com/urbit/urbit/issues/4613 are (I've concluded) coming out of this exception handler, and I think we can agree that these should not shut down the server.

Because of the (unfortunate) nature of the Haskell exception mechanism, it's not possible to confidently get a stable list of all the exceptions that might trip this handler, let alone fix a subset of those such that we can be sure its complement contains only things worth shutting down the server over. I'm just not comfortable implementing a policy like you propose and putting it on zod. I'd have to read all of the upstream libraries to find all places where exceptions are thrown, and I'd have to be confident that that doesn't change over time.

I'm going to issue a change here which uses the logger instead of stderr for printing exceptions, and maybe adds more info. (That's a good fix for the linked issue.)